### PR TITLE
Implement Residual block ABC, and concrete dilated residual block from randla-net

### PR DESCRIPTION
**Pull request by tristanheywood**
- Implement BaseResnetBlock in core modules, to act as the baseclass for the residual blocks from randla-net and KP conv
- move the random sampling from randlaconv to an ABC called BaseKNNConvolution
- Implement the dilated residual block from randla-net
- Implement RandLANetRes using this block - only 2 layers instead of the 4 layers from the paper. 

RandLANetRes achieves 89% training accuracy after half an epoch. 

I'm not sure about my method for dealing with the index when the convolution downsamples the pointcloud. At the moment I return an index if the pointcloud is down samples, otherwise None. It would be cleaner to always return an index (but less efficient).

### CODE REVIEW

File: models/core_modules.py [hunk 1]
```diff
@@ -62,6 +63,89 @@ def forward(self, data):
...
+
+    def forward(self, data):
+        x, pos, batch = data
+
+        if self.ratio == 1: #convolve every point
+            row, col = knn(pos, pos, self.k, batch, batch)
+            edge_index = torch.stack([col, row], dim=0)
+            x = self.conv(x, (pos, pos), edge_index)
+            return x, pos, batch, None
+        else: #downsample using self.sampling_strategy and convolve
```
Comment (nicolas-chaulet) [inline models/core_modules.py]: This else is not needed.  It will make the code more readable

File: models/core_modules.py [hunk 2]
```diff
@@ -62,6 +63,89 @@ def forward(self, data):
         data = (x, pos, batch)
         return data
 
+class BaseKNNConvolution(ABC, torch.nn.Module):
```
Comment (nicolas-chaulet) [inline models/core_modules.py]: A simpler solution would be to make this two helper functions that only do the sampling and not the convolution as opposed to a class, what do you think? Inheritance should be used with care

File: models/core_modules.py [hunk 3]
```diff
@@ -62,6 +63,89 @@ def forward(self, data):
...
+        self.activation = ReLU()
+
+    @abstractmethod
+    def convolution(self, data):
+        pass
+
+    def forward(self, data):
+
+        x, pos, batch = data #(N, indim)
+
```
Comment (nicolas-chaulet) [inline models/core_modules.py]: too many newlines here, Could you make the code a little more compact please?

File: models/RandLANet/modules.py [hunk 4]
```diff
@@ -14,24 +14,17 @@ class RandlaConv(MessagePassing):
...
-        x, pos, batch = data
-        idx = torch.randint(0, pos.shape[0], (math.floor(pos.shape[0]*self.ratio),))
-        row, col = knn(pos, pos[idx], self.k, batch, batch[idx])
-        edge_index = torch.stack([col, row], dim=0)
-        x = self.propagate(edge_index, x=x, pos=(pos, pos[idx]))
-
-        pos, batch = pos[idx], batch[idx]
-        data = (x, pos, batch)
-        return data 
+    def conv(self, x, pos, edge_index):
```
Comment (nicolas-chaulet) [inline models/RandLANet/modules.py]: Please see my comment rgarding changing BaseKNNConvolution to functions. Then The RandlaConv would still have a forward method.

**Comment by tchaton** [comment]
One small problem to solve with BaseResnetBlock. Its subConvolution needs idx being returned as data = (x, pos, batch, idx).
Normal convolution doesn't.
Some minors modifications (make to launch training on all model to be sure nothing is broken).
The API of BaseConvolution was changed and broke for other models.

**Comment by tristanheywood** [comment]
Randla-net with residual blocks gets 80% train accuracy after 60 iterations - looks like it's working

**Comment by tristanheywood** [comment]
I think we should just make all convolutions return the index.

File: models/KPConv/modules.py [hunk 5]
```diff
@@ -145,8 +145,11 @@ def __repr__(self):
 
 class KPConv(BaseConvolution):
     def __init__(self, ratio=None, radius=None, down_conv_nn=None, num_points=16, *args, **kwargs):
-        super(KPConv, self).__init__(ratio, radius)
+        super(KPConv, self).__init__(FPSSampler(ratio), RadiusNeighbourFinder(radius), *args, **kwargs)
 
+        self.ratio = ratio
```
Comment (nicolas-chaulet) [inline models/KPConv/modules.py]: These two properties are not used are they? If not please remove.
